server/src/main/java/io/crate/statistics/FetchSampleRequest.java

```diff
@@ -74,4 +74,24 @@ public List<Reference> columns() {
     public int maxSamples() {
         return maxSamples;
     }
+
+    @Override
+    public final int hashCode() {
+        // We fetch each relation in dedicated request, there won't be 2 requests for the same relation.
+        return relationName.hashCode();
+    }
+
+    @Override
+    public final boolean equals(Object obj) {
+        if (this == obj) {
+            return true;
+        }
+        if (obj == null) {
+            return false;
+        }
+        if (!(obj instanceof FetchSampleRequest other)) {
+            return false;
+        }
+        return relationName.equals(other.relationName);
+    }
```

Comment:
Why not include all parameters of the request? I think it's a bit dangerous to only take `relationName` into account.
E.g. right now it could be already possible that 2 requests will arrive for the same relation but with different columns as the table schema changed in-between. I think this would screw the result receiver...

Reply:
Good catch, thx. Pushed a fixup - there are still some things which can be improved but not sure:
- included `maxSamples` to `hashCode/equals` even though it's always a constant - I think if we pass it we might change it, otherwise could always use constant on the receiving side.
- currently all `Reference` implementations override `hashCode/equals`. Maybe we could add default implementation to `Reference` throwing an exception like "iml-n should override equals/hashCode" to avoid situation where we have 2 different `SomeNewRefImpl`  objects being not equal by ref but equal by internal props.

Reply:
> included maxSamples to hashCode/equals even though it's always a constant currently - I think if we pass it we might change it, otherwise could always use constant on the receiving side.

Yes valid point, I'm not sure why we stream it if it's hardcoded, maybe @mfussenegger could elaborate on this.

> currently all Reference implementations override hashCode/equals. Maybe we could add default implementation to Reference throwing an exception like "iml-n should override equals/hashCode" to avoid situation where we have 2 different SomeNewRefImpl objects being not equal by ref but equal by internal props.

👍 good point

Reply:
on refs:

```
 @Override
 default boolean equals(Object o) 
```
 
doesn't compile with error `"default method 'equals' overrides a member of 'java.lang.Object'",  looks like cannot do that on interface.

on constant:

we can stop streaming it and don't read/use hardcoded value on receiver with some version checks. Not sure it worth it  - AFAIK in total each `ANALYZE` fires `num_nodes * num_tables` requests. But let's see what others say and can follow up on this separately.